xlators/meta/src/options-dir.c

```diff
@@ -51,7 +51,72 @@ options_dir_fill(xlator_t *this, inode_t *inode, struct meta_dirent **dp)
     return xl->options->count;
 }
 
+static int
+adjustable_options_dir_fill(xlator_t *this, inode_t *inode,
+                            struct meta_dirent **dp)
+{
+    int count = 0;
+    int i = 0;
+    volume_opt_list_t *vol_opt_list = NULL;
+    volume_option_t *vol_opt = NULL;
+    struct meta_dirent *dirent = NULL;
+    struct meta_dirent *direntp = NULL;
+    xlator_t *xl = NULL;
+
+    xl = meta_ctx_get(inode, this);
+
+    list_for_each_entry(vol_opt_list, &xl->volume_options, list)
+    {
+        vol_opt = vol_opt_list->given_opt;
+        while (vol_opt->key[0]) {
+            if (vol_opt->fetch) {
+                for (i = 0; i < ZR_VOLUME_MAX_NUM_KEY; i++) {
+                    if (vol_opt->key[i])
+                        count++;
+                    else
+                        break;
+                }
+            }
+            vol_opt++;
+        }
+    }
+    dirent = GF_CALLOC(sizeof(*dirent), count, gf_meta_mt_dirents_t);
+    if (!dirent)
+        return -1;
+
+    direntp = dirent;
+
+    list_for_each_entry(vol_opt_list, &xl->volume_options, list)
+    {
+        vol_opt = vol_opt_list->given_opt;
+        while (vol_opt->key[0]) {
+            if (vol_opt->fetch) {
+                direntp->name = gf_strdup(vol_opt->key[0]);
```

Comment:
Yes, `diirentp->name` needs to be dynamically allocated, because it gets freed (in `meta_fd_release` and `meta_default_lookup`). Adding null check with discussion.